kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+# KEP - Real Load Aware Scheduling
+
+
+## Summary
+
+Minimizing machine costs by utilizing all nodes is the main objective for efficient cluster management. To achieve this goal, we can make the Kubernetes scheduler aware of the gap between resource allocation and actual resource utilization. Taking advantage of the gap may help pack pods more efficiently, while the default scheduler that only considers pod requests and allocable resources on nodes cannot.
+
+## Motivation
+
+Kubernetes provides a declarative resource model that core components (scheduler and kubelet) honor to behave consistently and satisfy QoS guarantees.
+However, using this model can lead to low-utilization of the cluster for the following reasons:
+
+1. It is hard for users to estimate accurate resource usage for applications. Also, users may not understand the resource model and not set it at all.
+2. The default in-tree scheduling plugins (Score) that Kubernetes provides don't consider live node utilization values.
```

Comment:
Playing devils' advocate here: Have you considered using the Vertical Pod Autoscaler and the existing plugins?

Reply:
My option: VPA may become more practical when in-place pod req update is supported in the upstream. For now, every adjustment VPA makes needs to restart the Pod (correct me if I'm wrong), which is somewhat unacceptable is most scenarios.

Reply:
Yeah, in place updates was what was in my mind.
It doesn't look like there has been any progress on that :/ https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Reply:
Yes, I was trying to help with the in-place pod update KEP. As it required some API changes, it may take a longer time to merge upstream. We need to wait at least until 1.20.

Besides, VPA with in-place update cannot solve all the goals we would like to achieve. 
- For example, the risk balancing plugin is taking the usage variations into consideration. Without understanding how the actual usage varies over time, it is impossible to estimate the risk. 
- The BFBP plugin aims to maintain all nodes' average utilization of around x%. VPA applied to a group of pods may have more accurate requests, but they may not help in achieving the overall node utilization goal in a cluster. Even if VPA is applied for all pods in a cluster, taking the average load on nodes in scoring nodes is much easier than taking the requests as they might be fluctuating all the time.